qCC/ccCommandLineCommands.cpp

```diff
@@ -1318,22 +1429,117 @@ bool CommandSubsample::process(ccCommandLineInterface& cmd)
 		{
+			CCCoreLib::ReferenceCloud* refCloud;
+			ccPointCloud* result;
+			unsigned sizeOfInputCloud = desc.pc->size();
+
 			cmd.print(QObject::tr("\tProcessing cloud %1").arg(!desc.pc->getName().isEmpty() ? desc.pc->getName() : "no name"));
-			
-			CCCoreLib::ReferenceCloud* refCloud = CCCoreLib::CloudSamplingTools::subsampleCloudWithOctreeAtLevel(	desc.pc,
-																													static_cast<unsigned char>(octreeLevel),
-																													CCCoreLib::CloudSamplingTools::NEAREST_POINT_TO_CELL_CENTER,
-																													progressDialog.data());
-			if (!refCloud)
+
+			if (byCellSize)
 			{
-				return cmd.error(QObject::tr("Subsampling process failed!"));
+				//calculate OCTREE level for each cloud based on required octree cell size
+				octreeLevel = ceil(log(desc.pc->getOwnBB().getMaxBoxDim() / cellSize) / log(2));
 			}
-			cmd.print(QObject::tr("\tResult: %1 points").arg(refCloud->size()));
-			
-			//save output
-			ccPointCloud* result = desc.pc->partialClone(refCloud);
-			delete refCloud;
-			refCloud = nullptr;
+
+			if (byMaxNumberOfPoints)
+			{
+				if (isPercent)
+				{
+					maxNumberOfPoints = ceil(sizeOfInputCloud * percent / 100);
+					cmd.print(QObject::tr("\tOutput point target: %1 * %2% = %3").arg(sizeOfInputCloud).arg(percent).arg(maxNumberOfPoints));
+				}
+				//calculate OCTREE level for each cloud based on required number of points
+				octreeLevel = ceil(log(maxNumberOfPoints) / (3.0 * log(2)));
+			}
+
+			//only process further if CELL_SIZE or octree level was given, or the numberOfPoints smaller than the input cloud
+			if (!byMaxNumberOfPoints || (byMaxNumberOfPoints && sizeOfInputCloud > maxNumberOfPoints))
+			{
+				//overwrite and print out finalized OCTREE level.
+				if (byCellSize || byMaxNumberOfPoints)
+				{
+					if (octreeLevel < 1)
+					{
+						//overwrite with min value instead of throwing an error
+						octreeLevel = 1;
+					}
+					if (octreeLevel > CCCoreLib::DgmOctree::MAX_OCTREE_LEVEL)
+					{
+						//overwrite with max value instead of throwing an error
+						octreeLevel = CCCoreLib::DgmOctree::MAX_OCTREE_LEVEL;
+					}
+					cmd.print(QObject::tr("\tCalculated octree level: %1").arg(octreeLevel));
+				}
+				refCloud = CCCoreLib::CloudSamplingTools::subsampleCloudWithOctreeAtLevel(	desc.pc,
+																							static_cast<unsigned char>(octreeLevel),
+																							CCCoreLib::CloudSamplingTools::NEAREST_POINT_TO_CELL_CENTER,
+																							progressDialog.data());
+				CCCoreLib::ReferenceCloud* refCloud2;
+				//predict and calculate new octree levels to match target number only if NUMBER_OF_POINTS is defined
+				while (byMaxNumberOfPoints && refCloud->size() < maxNumberOfPoints && octreeLevel < CCCoreLib::DgmOctree::MAX_OCTREE_LEVEL)
+				{
+					//predict a new octree lvl based on new informations, it is faster then go through one by one. And it is predicting on the safe side, will not overshoot at any circumstance
```

Comment:
```suggestion
					//predict a new octree level based on new information. It is faster than going through one by one. And it is predicting on the safe side. It will not overshoot in any circumstances.
```

Reply:
And frankly, I doubt that's true on large clouds (also for the sake of memory).

Just compute the octree before calling CCCoreLib::CloudSamplingTools::subsampleCloudWithOctreeAtLevel, and then:
- either you can use DgmOctree::findBestLevelForAGivenCellNumber, and if you overshoot, you just take the next level
- or you can use DgmOctree::getCellNumber that will give you the number of cells/points per level (see https://github.com/CloudCompare/CCCoreLib/blob/master/include/DgmOctree.h#L946)